### PR TITLE
Fix caching issue with reversed ranges

### LINKED ISSUE #1106: `Incorrect array size` error for some cases of reversed ranges with absolute addressing (previously: Partially locked, reversed range breaks setCellContents)
### Description

setCellContents on hyperformula instance with the following data causes an "Incorrect array size" error. 

```
const tableData = [
  [1, "=SUM(A1:A$1)"],
  [1, "=SUM(A2:A$1)"]
];
```

### Demo
https://codesandbox.io/s/frosty-meadow-umnbk2?file=/src/data.js

Partially locked ranges that are in the correct order work.
Calculating a single reversed and locked formula seems to work fine. See this repo here: https://codesandbox.io/s/silly-leftpad-d38lgu?file=/src/data.js

HyperFormula version: 2.2

### Links
Related #835

### PR DESCRIPTION
**Pull request by sequba**
### Context

This bug was quite complicated. It involved using 2 ranges, which
- have the same R1C1 representation
- evaluate to different values
- exactly one of them is reversed

The reversed ranges are handled by converting them to regular ranges. Before this PR, this conversion was performed during the parsing stage. The AST of a reversed range is different before and after the conversion, but our parser uses caching and outputs the same ASTs for formulas that have the same R1C1 representation. In this PR I moved the code that handles reversed ranges to be run in the post-parsing stage so it is independent of the caching mechanism of the parser.

### How did you test your changes?

- [x] unit tests pass
- [x] add unit tests for more corner cases

### Types of changes

- [ ] Breaking change (a fix or a feature because of which an existing functionality doesn't work as expected anymore)
- [ ] New feature or improvement (a non-breaking change that adds functionality)
- [x] Bug fix (a non-breaking change that fixes an issue)
- [ ] Additional language file, or a change to an existing language file (translations)
- [ ] Change to the documentation

### Related issues:
Fixes #1106 

### Checklist:

- [x] My change is compliant with the [OpenDocument](https://docs.oasis-open.org/office/OpenDocument/v1.3/os/part4-formula/OpenDocument-v1.3-os-part4-formula.html) standard.
- [x] My change is compatible with Microsoft Excel.
- [x] My change is compatible with Google Sheets.
- [x] My code follows the code style of this project.
- [x] I described my changes in the [CHANGELOG.md](https://github.com/handsontable/hyperformula/blob/master/CHANGELOG.md) file.
- [ ] My changes require a documentation update.
- [ ] My changes require a migration guide.

### CODE REVIEW

**Review by adamu-handsontable — COMMENTED** [review]
I've added some questions regarding this PR and proposed more tests.

File: test/AbsoluteCellRange.spec.ts [hunk 1]
```diff
@@ -34,5 +33,25 @@ describe('AbsoluteCellRange', () => {
...
+        const start = { sheet: 0, row: 42, col: 42 }
+        const end = { sheet: 0, row: 666, col: 666 }
+        const range = AbsoluteCellRange.fromSimpleCellAddresses(start, end)
+
+        expect(range).toBeInstanceOf(AbsoluteCellRange)
+        expect(range).not.toBeInstanceOf(AbsoluteColumnRange)
+        expect(range).not.toBeInstanceOf(AbsoluteRowRange)
+      })
+
+      it('constructs a AbsoluteColumnRange when called with end.row = Infinity', () => {
```
[the remark this answers is not in the document]
Comment (sequba) [reply test/AbsoluteCellRange.spec.ts:50]: added

File: src/AbsoluteCellRange.ts [hunk 2]
```diff
@@ -55,6 +55,21 @@ export class AbsoluteCellRange implements SimpleCellRange {
...
+
+    if (Number.isFinite(height) && Number.isFinite(width)) {
+      return new AbsoluteCellRange(start, end)
+    }
+
+    if (Number.isFinite(height)) {
+      return new AbsoluteRowRange(start.sheet, start.row, end.row)
+    }
+
+    return new AbsoluteColumnRange(start.sheet, start.col, end.col)
```
[the remark this answers is not in the document]
Comment (sequba) [reply src/AbsoluteCellRange.ts:70]: Well, in theory, it will be handled by this factory method. It added a unit test for that as suggested in another comment. But keep in mind, that such a range will never be created because the formula syntax does not allow it. There are only cell ranges (`A1:D4`), column ranges (`A:D`) and row ranges (`1:4`).

File: src/AbsoluteCellRange.ts [hunk 3]
```diff
@@ -55,6 +55,21 @@ export class AbsoluteCellRange implements SimpleCellRange {
     return this.start.sheet
   }
 
+  public static fromSimpleCellAddresses(start: SimpleCellAddress, end: SimpleCellAddress): AbsoluteCellRange {
+    const width = end.col - start.col
```
[the remark this answers is not in the document]
Comment (sequba) [reply src/AbsoluteCellRange.ts:63]: Good point. It totally makes sense to check for 3D ranges. I added the check and unit tests for this case.

File: src/parser/ParserWithCaching.ts [hunk 4]
```diff
@@ -97,9 +101,114 @@ export class ParserWithCaching {
...
 
-    return {ast, errors: [], hasVolatileFunction, hasStructuralChangeFunction, dependencies: relativeDependencies}
+    const {ast, hasVolatileFunction, hasStructuralChangeFunction } = cacheResult
+    const astWithNoReversedRanges = this.convertReversedRangesToRegularRanges(ast)
+    const dependencies = collectDependencies(astWithNoReversedRanges, this.functionRegistry)
+
+    return {ast: astWithNoReversedRanges, errors: [], hasVolatileFunction, hasStructuralChangeFunction, dependencies }
+  }
+
+  private convertReversedRangesToRegularRanges(ast: Ast): Ast {
```
[the remark this answers is not in the document]
Comment (sequba) [reply src/parser/ParserWithCaching.ts:112]: Actually, I thought about using the [Visitor design pattern](https://refactoring.guru/design-patterns/visitor) which is suitable for algorithms that traverse tree-like structures. But that would be a major refactor indeed. I don't want to delay the release so WDYT about creating an issue for the next release to do such an improbement?

**Review by adamu-handsontable — APPROVED** [review]
Everything looks good. Nice job, thank you

File: src/parser/ParserWithCaching.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (sequba) [reply src/parser/ParserWithCaching.ts:112]: https://github.com/handsontable/hyperformula/issues/1169
